Index: java/src/com/tigervnc/vncviewer/MANIFEST.MF
===================================================================
--- java/src/com/tigervnc/vncviewer/MANIFEST.MF	(revision 4001)
+++ java/src/com/tigervnc/vncviewer/MANIFEST.MF	(revision 4002)
@@ -1,2 +1,2 @@
 Manifest-Version: 1.0
-Main-Class: VncViewer
+Main-Class: com.tigervnc.vncviewer.VncViewer
Index: java/src/com/tigervnc/vncviewer/index.html
===================================================================
--- java/src/com/tigervnc/vncviewer/index.html	(revision 4001)
+++ java/src/com/tigervnc/vncviewer/index.html	(revision 4002)
@@ -20,7 +20,7 @@
 <TITLE>
 TigerVNC desktop
 </TITLE>
-<APPLET CODE="VncViewer.class" ARCHIVE="VncViewer.jar"
+<APPLET CODE="com.tigervnc.vncviewer.VncViewer" ARCHIVE="VncViewer.jar"
         WIDTH="800" HEIGHT="632">
 <PARAM NAME="PORT" VALUE="5901">
 </APPLET>
Index: java/src/com/tigervnc/vncviewer/Makefile
===================================================================
--- java/src/com/tigervnc/vncviewer/Makefile	(revision 4001)
+++ java/src/com/tigervnc/vncviewer/Makefile	(revision 4002)
@@ -35,7 +35,12 @@
 	$(JC) $(JCFLAGS) -O $(SOURCES)
 
 $(ARCHIVE): $(CLASSES) $(MANIFEST)
-	$(JAR) cfm $(ARCHIVE) $(MANIFEST) $(CLASSES)
+	cd ../../..; \
+	$(JAR) cfm com/tigervnc/vncviewer/$(ARCHIVE) \
+		com/tigervnc/vncviewer/$(MANIFEST) \
+		com/tigervnc/vncviewer/*.class \
+		com/tigervnc/decoder/*.class \
+		com/tigervnc/decoder/common/*.class
 
 install: $(CLASSES) $(ARCHIVE)
 	$(CP) $(CLASSES) $(ARCHIVE) $(PAGES) $(INSTALL_DIR)
@@ -44,4 +49,4 @@
 	@$(ExportJavaClasses)
 
 clean::
-	$(RM) *.class *.jar
+	$(RM) *.class *.jar ../decoder/*.class ../decoder/common/*.class
