Index: unix/xserver/hw/vnc/xf86vncModule.cc
===================================================================
--- unix/xserver/hw/vnc/xf86vncModule.cc	(revision 4004)
+++ unix/xserver/hw/vnc/xf86vncModule.cc	(revision 4005)
@@ -43,6 +43,8 @@
 extern void vncExtensionInit();
 static void vncExtensionInitWithParams(INITARGS);
 
+char *listenaddr = NULL;
+
 static MODULESETUPPROTO(vncSetup);
 
 ExtensionModule vncExt =
